File: examples/loop-monitoring.js
import { DashClaw } from '../sdk/dashclaw.js';
import dotenv from 'dotenv';
dotenv.config();

const claw = new DashClaw({
  baseUrl: process.env.DASHCLAW_BASE_URL || 'http://localhost:3000',
  apiKey: process.env.DASHCLAW_API_KEY,
  agentId: 'node-loop-agent',
});

async function main() {
  console.log('Starting background loop...');

  // 1. Create a parent action for the loop to attach to
  const { action } = await claw.createAction({
    action_type: 'indexing',
    declared_goal: 'Index new documentation for RAG',
    risk_score: 15
  });

  const actionId = action.action_id;
  console.log(`Action created: ${actionId}`);

  // 2. Register an open loop against the action
  const { loop_id } = await claw.registerOpenLoop(
    actionId,
    'background_indexing',
    'Indexing new documentation for RAG',
    { priority: 'medium' }
  );

  console.log(`Loop created: ${loop_id}`);

  // 3. Simulate work
  await new Promise(r => setTimeout(r, 1000));
  console.log('Processed 15 files...');

  // 4. Resolve the loop
  await new Promise(r => setTimeout(r, 1000));
  await claw.resolveOpenLoop(loop_id, 'resolved', 'Successfully indexed 42 files.');

  console.log('Loop resolved.');

  // 5. Close out the parent action
  await claw.updateOutcome(actionId, {
    status: 'completed',
    output_summary: 'Indexed 42 files for RAG.'
  });

  console.log('Action complete.');
}

main();

File: examples/loop-monitoring.py
import os
import time
from dashclaw import DashClaw
from dotenv import load_dotenv

load_dotenv()

claw = DashClaw(
    base_url=os.environ.get("DASHCLAW_BASE_URL", "http://localhost:3000"),
    api_key=os.environ["DASHCLAW_API_KEY"],
    agent_id="loop-monitor-bot",
)

def main():
    # 1. Create a parent action for the loop to attach to
    action = claw.create_action(
        action_type="security_scan",
        declared_goal="Continuous vulnerability scan of network subnet",
        risk_score=35,
    )
    action_id = action.get("action_id") or action.get("action", {}).get("action_id")
    print(f"Action created: {action_id}")

    # 2. Register an open loop against the action
    loop = claw.register_open_loop(
        action_id=action_id,
        loop_type="security_scan",
        description="Continuous vulnerability scan of network subnet",
        priority="high",
        owner="security-bot-1",
    )
    loop_id = loop.get("loop_id")
    print(f"Loop created: {loop_id}")

    # 3. Simulate progress
    time.sleep(2)
    print("Scan 50% complete. No critical issues found.")

    # 4. Resolve the loop
    time.sleep(2)
    claw.resolve_open_loop(
        loop_id,
        status="resolved",
        resolution="Full subnet scan completed. 3 low-severity findings identified.",
    )
    print(f"Loop {loop_id} resolved.")

    # 5. Close out the parent action
    claw.update_outcome(action_id, status="completed",
                        output_summary="Security scan complete. 3 low-severity findings.")
    print("Action complete.")

if __name__ == "__main__":
    main()
